### PR TITLE
test: Run symlink regression tests on Windows

### PR DESCRIPTION
**Pull request by maflcko**
Seems odd to add tests, but not run them on the platform that needs them most.

### CODE REVIEW

**Comment by vasild** [comment]
Hmm, symlinks don't exist on Windows (or more precisely on Windows-supported filesystems like FAT16, FAT32, NTFS, whatnot?).

**Comment by prayank23** [comment]
> Hmm, symlinks don't exist on Windows (or more precisely on Windows-supported filesystems like FAT16, FAT32, NTFS, whatnot?).

_The Windows’ NTFS file system has supported symlinks since Windows Vista._

https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/

**Comment by MarcoFalke** [comment]
Symlinks exist on Windows, otherwise the test wouldn't pass (https://github.com/bitcoin/bitcoin/pull/24338/files#r808801652).

Though, it looks like symlinks don't exist on mingw: https://gcc.gnu.org/git/?p=gcc.git;a=blobdiff;f=libstdc%2B%2B-v3/testsuite/27_io/filesystem/operations/create_directories.cc;h=304c1453afe650b1e5850114df4d86772d6eb4f4;hp=393d6a5530925bd1305eaf460c107954cb0b1ab9;hb=124eaa50e0a34f5f89572c1aa812c50979da58fc;hpb=e07d30fdcaec4906e0dcb948fc4748bf74c15c05

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa1919014c23ef8f80e7893e1a4cee9bfa375eb9

It's an anti-pattern to skip tests on windows, and this has caused bugs recently. We should almost never be skipping tests on windows. If something is expected to fail on windows, the test should run and then check for the expected failure.

**Comment by vasild** [comment]
Does the test pass only if the underlying filesystem is NTFS? Would it fail on a FAT filesystem?

**Comment by MarcoFalke** [comment]
Looks like the new test uncovered a new libc++ bug on Bionic? Removed new test for now. I'll debug this as a follow-up.

Should be trivial to re-ACK.

**Review by ryanofsky — APPROVED** [review]
Code review ACK fad7ddf9e3710405d727f61d8200d5efed1e705b, just removing new test. Would be nice if the test could be added later, of course.

> Does the test pass only if the underlying filesystem is NTFS? Would it fail on a FAT filesystem?

Probably trying to create symlinks on a fat drive would fail whether the test is running on windows or linux. Maybe it'd be useful to have CI build that runs tests on a FAT filesystem (if this is something we support). I'd be surprised if the tests already pass already and there wouldn't be other bugs to fix.

**Comment by MarcoFalke** [comment]
The multiwallet test already uses symlinks unconditionally, so this pull isn't changing where test can be run. Is FAT still a thing in 2022? If yes, it might be better to start a new issue/thread to discuss testing on it.

**Comment by luke-jr** [comment]
Aside from FAT, Android also prohibits symlinks on the /sdcard filesystem

```
:/sdcard # ln -s a b
ln: cannot create symbolic link from 'a' to 'b': Operation not permitted
```

**Review by ghost — COMMENTED** [review]
Concept ACK

> Aside from FAT, Android also prohibits symlinks on the /sdcard filesystem
:/sdcard # ln -s a b

@luke-jr this works for me because file system is ext4. Symlinks work on Android, it's just FAT file system that does not support it.

File: src/test/fs_tests.cpp [hunk 1]
```diff
@@ -152,7 +152,7 @@ BOOST_AUTO_TEST_CASE(rename)
     fs::remove(path2);
 }
 
-#ifndef WIN32
+#ifndef __MINGW64__ // no symlinks on mingw
```
Comment (hebasto) [inline src/test/fs_tests.cpp:155]: Also see https://github.com/bitcoin/bitcoin/pull/24338/files#r808540886

**Review by hebasto — COMMENTED** [review]
Concept ACK.

**Comment by luke-jr** [comment]
>@luke-jr this works for me because file system is ext4. Symlinks work on Android, it's just FAT file system that does not support it.

On Android, /sdcard is "sdcardfs", which doesn't support symlinks.

File: test/functional/feature_dirsymlinks.py [hunk 2]
```diff
@@ -6,34 +6,36 @@
...
-    def skip_test_if_missing_module(self):
-        if sys.platform == 'win32':
-            raise SkipTest("Symlinks test skipped on Windows")
 
+class SymlinkTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 1
 
     def run_test(self):
+        dir_new_blocks = self.nodes[0].chain_path / "new_blocks"
```
Comment (laanwj) [inline test/functional/feature_dirsymlinks.py:24]: Any specific reason to assign these to variables?

File: test/functional/feature_dirsymlinks.py [hunk 3]
```diff
@@ -6,34 +6,36 @@
...
+        rename_and_link(
+            from_name=self.nodes[0].chain_path / "chainstate",
+            to_name=dir_new_chainstate,
+        )
 
         self.start_node(0)
 
 
-if __name__ == '__main__':
+if __name__ == "__main__":
```
Comment (laanwj) [inline test/functional/feature_dirsymlinks.py:40]: Why change the quote style here? Looking at the current `.py` files the syntax here varies wildly. I mean, I'm not against making it consistent, but changing just this file seems pointless.

**Comment by laanwj** [comment]
> It's an anti-pattern to skip tests on windows, and this has caused bugs recently. 

Look, I had no idea symlinks were a reliable feature on Windows (they were not for a long time) and am unable to test anything on Windows myself. This test was added to test a workaround for a bug in std::fs on gcc/glibc. Disabling it for windows seemed to make sense. I'm fine with enabling it, but I don't think I committed some anti-patttern sin by doing it that way.

File: test/functional/feature_dirsymlinks.py (hunk 2 above)
Comment (MarcoFalke) [reply test/functional/feature_dirsymlinks.py:24]: Sorry, no reason. I needed this for the follow-up test. Though, the test uncovered another bug (https://github.com/bitcoin/bitcoin/pull/24381#issuecomment-1044598093), so I removed it.

Happy to revert the variables as well. Lmk?

File: test/functional/feature_dirsymlinks.py (hunk 3 above)
Comment (MarcoFalke) [reply test/functional/feature_dirsymlinks.py:40]: No reason. I think my editor/formatter did this. Happy to revert.

Comment (laanwj) [reply test/functional/feature_dirsymlinks.py:40]: No strong opinion on reverting. I was just wondering.

File: test/functional/feature_dirsymlinks.py (hunk 2 above)
Comment (laanwj) [reply test/functional/feature_dirsymlinks.py:24]: No, if you need it for a follow-up test it's fine.

**Comment by ryanofsky** [comment]
Nope, laanwj you have sinned and I pronounce you guilty of crimes against test coverage! But seriously, if you want feature X to work on all platforms and you disable tests for feature X on certain platforms, it's just a red flag. You don't have to go to jail, or devote yourself to fixing the test when there are more important things to do, and my comment was motivated more by #24251 than the create_directories thing. I just think skipping tests without a very specific and definitive reason is something people should be suspicious of.

**Comment by laanwj** [comment]
LOL :smile: 

Code review ACK fad7ddf9e3710405d727f61d8200d5efed1e705b

**Pull request merged**